File: studio/services/profile_service.py
"""Business profile persistence, validation, and generation gate."""

import re

from studio.repositories import profiles_repo, projects_repo

_REQUIRED_FIELDS = [
    ("brand.brandName", "brand name"),
    ("business.type", "business type"),
    ("business.services", "at least one service"),
    ("business.targetAudience", "target audience"),
    ("contact.email", "contact email"),
    ("contact.phone", "contact phone"),
]

_EMAIL_RE = re.compile(r"^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$", re.I)
_PHONE_RE = re.compile(r"\+?\d[\d\s().-]{7,}\d")

_MERGE_KEYS = frozenset({"brand", "business", "location", "contact"})


class ProfileValidationError(ValueError):
    """Raised when a profile patch fails validation."""


class ProfileIncompleteError(Exception):
    """Raised when generation is requested before the minimum profile gate passes."""

    def __init__(self, missing_fields):
        self.missing_fields = list(missing_fields)
        super().__init__("profile incomplete")


def _get_nested(profile, path):
    cur = profile
    for part in path.split("."):
        cur = cur.get(part) if isinstance(cur, dict) else None
        if cur is None:
            return None
    return cur


def _missing_fields(profile):
    missing = []
    for path, label in _REQUIRED_FIELDS:
        val = _get_nested(profile, path)
        if path == "business.services":
            if not val:
                missing.append(label)
        elif not val:
            missing.append(label)
    return missing


def compute_completeness(profile):
    total = len(_REQUIRED_FIELDS)
    missing = _missing_fields(profile)
    done = total - len(missing)
    percent = int(round((done / total) * 100)) if total else 0
    profile["completeness"] = {"percent": percent, "missingFields": missing}
    return profile["completeness"]


def evaluate_generation_gate(profile):
    completeness = compute_completeness(profile)
    return {
        "ready": not completeness["missingFields"],
        "percent": completeness["percent"],
        "missingFields": completeness["missingFields"],
    }


def _deep_merge(base, patch):
    for key, value in patch.items():
        if key not in _MERGE_KEYS:
            continue
        if isinstance(value, dict) and isinstance(base.get(key), dict):
            merged = dict(base[key])
            merged.update(value)
            base[key] = merged
        else:
            base[key] = value
    return base


def _validate_profile(profile):
    email = _get_nested(profile, "contact.email")
    if email and not _EMAIL_RE.match(str(email)):
        raise ProfileValidationError("invalid contact email")
    phone = _get_nested(profile, "contact.phone")
    if phone and not _PHONE_RE.search(str(phone)):
        raise ProfileValidationError("invalid contact phone")
    services = _get_nested(profile, "business.services")
    if services is not None and not isinstance(services, list):
        raise ProfileValidationError("business.services must be a list")


def get_profile(project_id):
    if not projects_repo.get(project_id):
        raise ValueError("project not found")
    profile = profiles_repo.get(project_id)
    gate = evaluate_generation_gate(profile)
    profile["gate"] = gate
    return profile


def update_profile(project_id, patch):
    if not projects_repo.get(project_id):
        raise ValueError("project not found")
    profile = profiles_repo.get(project_id)
    _deep_merge(profile, patch or {})
    _validate_profile(profile)
    compute_completeness(profile)
    profiles_repo.save(profile)
    gate = evaluate_generation_gate(profile)
    profile["gate"] = gate
    return profile


def require_generation_ready(project_id):
    profile = profiles_repo.get(project_id)
    gate = evaluate_generation_gate(profile)
    if not gate["ready"]:
        raise ProfileIncompleteError(gate["missingFields"])
    return profile

File: src/server/services/profileService.ts
import * as profilesRepo from "../repos/profilesRepo";
import * as projectsRepo from "../repos/projectsRepo";

const REQUIRED_FIELDS: Array<[string, string]> = [
  ["brand.brandName", "brand name"],
  ["business.type", "business type"],
  ["business.services", "at least one service"],
  ["business.description", "homepage intro line"],
  ["business.targetAudience", "target audience"],
  ["location.country", "country"],
  ["location.city", "city"],
];

const EMAIL_RE = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i;
const PHONE_RE = /\+?\d[\d\s().-]{7,}\d/;
const MULTI_PLACE_RE = /,|\band\b/i;
const MERGE_KEYS = new Set(["brand", "business", "location", "contact", "design", "skipped"]);

export class ProfileValidationError extends Error {}
export class ProfileIncompleteError extends Error {
  constructor(readonly missingFields: string[]) {
    super("profile incomplete");
  }
}

export function isMultiPlaceValue(value: unknown): boolean {
  if (!value || typeof value !== "string") return false;
  const parts = value.split(MULTI_PLACE_RE).map((p) => p.trim()).filter(Boolean);
  return parts.length >= 2;
}

export function splitMultiPlaceValue(value: unknown): string[] {
  if (!value || typeof value !== "string") return [];
  return value.split(MULTI_PLACE_RE).map((p) => p.trim()).filter(Boolean);
}

function getNested(profile: Record<string, unknown>, path: string): unknown {
  let cur: unknown = profile;
  for (const part of path.split(".")) {
    cur = typeof cur === "object" && cur !== null ? (cur as Record<string, unknown>)[part] : null;
    if (cur == null) return null;
  }
  return cur;
}

function missingFields(profile: Record<string, unknown>): string[] {
  const skipped = new Set((profile.skipped as string[] | undefined) ?? []);
  const missing: string[] = [];
  for (const [path, label] of REQUIRED_FIELDS) {
    if (skipped.has(path)) continue;
    const val = getNested(profile, path);
    if (path === "business.services") {
      if (!val || (Array.isArray(val) && val.length === 0)) missing.push(label);
    } else if (path === "location.city") {
      const country = getNested(profile, "location.country");
      if (isMultiPlaceValue(country)) continue;
      if (!val) missing.push(label);
    } else if (!val) {
      missing.push(label);
    }
  }
  return missing;
}

export function computeCompleteness(profile: Record<string, unknown>): Record<string, unknown> {
  const total = REQUIRED_FIELDS.length;
  const missing = missingFields(profile);
  const done = total - missing.length;
  const percent = total ? Math.round((done / total) * 100) : 0;
  profile.completeness = { percent, missingFields: missing };
  return profile.completeness as Record<string, unknown>;
}

export function evaluateGenerationGate(profile: Record<string, unknown>): Record<string, unknown> {
  const completeness = computeCompleteness(profile);
  return {
    ready: !(completeness.missingFields as string[]).length,
    percent: completeness.percent,
    missingFields: completeness.missingFields,
  };
}

function deepMerge(base: Record<string, unknown>, patch: Record<string, unknown>): Record<string, unknown> {
  for (const [key, value] of Object.entries(patch)) {
    if (!MERGE_KEYS.has(key)) continue;
    if (typeof value === "object" && value !== null && typeof base[key] === "object" && base[key] !== null) {
      base[key] = { ...(base[key] as Record<string, unknown>), ...(value as Record<string, unknown>) };
    } else {
      base[key] = value;
    }
  }
  return base;
}

function validateProfile(profile: Record<string, unknown>): void {
  const email = getNested(profile, "contact.email");
  if (email && !EMAIL_RE.test(String(email))) throw new ProfileValidationError("invalid contact email");
  const phone = getNested(profile, "contact.phone");
  if (phone && !PHONE_RE.test(String(phone))) throw new ProfileValidationError("invalid contact phone");
  const services = getNested(profile, "business.services");
  if (services != null && !Array.isArray(services)) {
    throw new ProfileValidationError("business.services must be a list");
  }
}

export async function getProfile(projectId: string): Promise<Record<string, unknown>> {
  if (!(await projectsRepo.get(projectId))) throw new Error("project not found");
  const profile = await profilesRepo.get(projectId);
  profile.gate = evaluateGenerationGate(profile);
  return profile;
}

export async function updateProfile(
  projectId: string,
  patch: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  if (!(await projectsRepo.get(projectId))) throw new Error("project not found");
  const profile = await profilesRepo.get(projectId);
  deepMerge(profile, patch ?? {});
  validateProfile(profile);
  computeCompleteness(profile);
  await profilesRepo.save(profile);
  profile.gate = evaluateGenerationGate(profile);
  return profile;
}

export async function requireGenerationReady(projectId: string): Promise<Record<string, unknown>> {
  const profile = await profilesRepo.get(projectId);
  const gate = evaluateGenerationGate(profile);
  if (!gate.ready) throw new ProfileIncompleteError(gate.missingFields as string[]);
  return profile;
}
